File: app/src/main/java/com/github/fo2rist/mclaren/ui/circuitsscreen/CircuitDetailsFragment.java
package com.github.fo2rist.mclaren.ui.circuitsscreen;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.fo2rist.mclaren.R;
import com.github.fo2rist.mclaren.ui.models.CalendarEvent;
import com.github.fo2rist.mclaren.ui.widgets.InformationLineView;
import com.github.fo2rist.mclaren.utils.IntentUtils;
import timber.log.Timber;

import static com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitDetailedImageUriById;
import static com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitDrsImageUriById;
import static com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitSectorsImageUriById;
import static com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitTurnsImageUriById;


public class CircuitDetailsFragment extends Fragment implements View.OnClickListener {

    private static final String ARG_EVENT = "event";

    @Nullable
    private CalendarEvent event;

    @NonNull
    public static Bundle createLaunchBundle(CalendarEvent event) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_EVENT, event);
        return args;
    }

    public static CircuitDetailsFragment newInstance(Bundle args) {
        CircuitDetailsFragment fragment = new CircuitDetailsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public CircuitDetailsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fetchBundleParameters();
    }

    private void fetchBundleParameters() {
        Bundle args = getArguments();
        if (args == null) {
            Timber.e("No arguments provided for fragment");
            //TODO forward error-level logging to Crashlytics. 2018-02-25
            return;
        }

        event = (CalendarEvent) args.getSerializable(ARG_EVENT);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_circuit_item, container, false);
        populateView(rootView);
        return rootView;
    }

    private void populateView(View rootView) {
        if (event == null) {
            Timber.e("No circuit model provided to display");
            return;
        }

        ((ImageView) rootView.findViewById(R.id.circuit_image))
                .setImageURI(getCircuitDetailedImageUriById(event.circuitId));
        ((ImageView) rootView.findViewById(R.id.circuit_drs_image))
                .setImageURI(getCircuitDrsImageUriById(event.circuitId));
        ((ImageView) rootView.findViewById(R.id.circuit_sectors_image))
                .setImageURI(getCircuitSectorsImageUriById(event.circuitId));
        ((ImageView) rootView.findViewById(R.id.circuit_turns_image))
                .setImageURI(getCircuitTurnsImageUriById(event.circuitId));

        TextView circuitTitleView = rootView.findViewById(R.id.circuit_title);
        circuitTitleView.setText(event.grandPrixName);
        circuitTitleView.setOnClickListener(this);

        TextView circuitDetailsView = rootView.findViewById(R.id.circuit_details);
        circuitDetailsView.setText(getString(R.string.circuit_details_format, event.city, event.trackName));

        LinearLayout propertiesList = rootView.findViewById(R.id.properties_linearlayout);

        addInformationLine(propertiesList, getString(R.string.circuit_details_laps), String.valueOf(event.laps));
        addInformationLine(propertiesList, getString(R.string.circuit_details_length), getString(R.string.distance_km_format, event.length));
        addInformationLine(propertiesList, getString(R.string.circuit_details_distance), getString(R.string.distance_km_format, event.distance));
        addInformationLine(propertiesList, getString(R.string.circuit_details_seasons), event.seasons);
        addInformationLine(propertiesList, getString(R.string.circuit_details_gp_held), String.valueOf(event.gpHeld));
    }

    private void addInformationLine(LinearLayout propertiesList, String propertyName, String propertyValue) {
        InformationLineView propertyView = new InformationLineView(getContext());
        propertyView.setContent(propertyName, propertyValue);
        propertiesList.addView(propertyView);
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.circuit_title) {
            IntentUtils.launchSafely(getContext(), IntentUtils.createBrowserIntent(event.wikiLink));
        }
    }
}

File: app/src/main/java/com/github/fo2rist/mclaren/ui/circuitsscreen/CircuitDetailsFragment.kt
package com.github.fo2rist.mclaren.ui.circuitsscreen

import android.os.Bundle
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.github.fo2rist.mclaren.R
import com.github.fo2rist.mclaren.databinding.ActivityCircuitDetailsBinding
import com.github.fo2rist.mclaren.databinding.FragmentCircuitItemBinding
import com.github.fo2rist.mclaren.ui.models.CalendarEvent
import com.github.fo2rist.mclaren.ui.widgets.InformationLineView
import com.github.fo2rist.mclaren.utils.IntentUtils
import com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitDetailedImageUriById
import com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitDrsImageUriById
import com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitSectorsImageUriById
import com.github.fo2rist.mclaren.utils.ResourcesUtils.getCircuitTurnsImageUriById

/**
 * Fragment that displays single Circuit's map and detailed info.
 */
class CircuitDetailsFragment : Fragment(), View.OnClickListener {

    private lateinit var binding: FragmentCircuitItemBinding

    private lateinit var event: CalendarEvent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fetchBundleParameters()
    }

    private fun fetchBundleParameters() {
        event = arguments?.getSerializable(ARG_EVENT) as? CalendarEvent
            ?: throw IllegalArgumentException("Required parameter `$ARG_EVENT` not present or not a CalendarEvent")
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCircuitItemBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        populateMapImage()
        populateMainInformation()
        populateDetailedInformationLines()
    }

    private fun populateMapImage() {
        binding.circuitImage.setImageURI(
                getCircuitDetailedImageUriById(event.circuitId))
        binding.circuitDrsImage.setImageURI(
                getCircuitDrsImageUriById(event.circuitId))
        binding.circuitSectorsImage.setImageURI(
                getCircuitSectorsImageUriById(event.circuitId))
        binding.circuitTurnsImage.setImageURI(
                getCircuitTurnsImageUriById(event.circuitId))
    }

    private fun populateMainInformation() {
        binding.circuitTitle.text = event.grandPrixName
        binding.circuitTitle.setOnClickListener(this)

        binding.circuitDetails.text = getString(R.string.circuit_details_format, event.city, event.trackName)
    }

    private fun populateDetailedInformationLines() {
        binding.propertiesLinearlayout.addInformationLine(
                R.string.circuit_details_laps,
                event.laps.toString())
        binding.propertiesLinearlayout.addInformationLine(
                R.string.circuit_details_length,
                getString(R.string.distance_km_format, event.length))
        binding.propertiesLinearlayout.addInformationLine(
                R.string.circuit_details_distance,
                getString(R.string.distance_km_format, event.distance))
        binding.propertiesLinearlayout.addInformationLine(
                R.string.circuit_details_seasons,
                event.seasons)
        binding.propertiesLinearlayout.addInformationLine(
                R.string.circuit_details_gp_held,
                event.gpHeld.toString())
    }

    private fun LinearLayout.addInformationLine(@StringRes propertyTitle: Int, propertyValue: String) {
        val propertyView = InformationLineView(context)
        propertyView.setContent(propertyTitle, propertyValue)
        addView(propertyView)
    }

    override fun onClick(view: View) {
        @SuppressWarnings("UnsafeCallOnNullableType")
        if (view.id == R.id.circuit_title) {
            IntentUtils.launchSafely(requireContext(), IntentUtils.createBrowserIntent(event.wikiLink))
        }
    }

    companion object {

        private const val ARG_EVENT = "event"

        /**
         * Create fragment to display circuit/event info.
         */
        @JvmStatic
        fun newInstance(event: CalendarEvent): CircuitDetailsFragment {
            return CircuitDetailsFragment().apply {
                arguments = Bundle().apply { putSerializable(ARG_EVENT, event) }
            }
        }
    }
}
